### DISCUSSION TITLE
Output free energy from GrainGrowthAction

### DISCUSSION THREAD

**Discussion by Clark-domain** [idaholab/moose#18518] [Q&A Modules: Phase field] [answered]
Hi, team,

I am trying to figure out how to use the GrainGrowthAction in the phase-field module to simulate the polycrystal growth behavior.  I have the problem to output the free energy.

In the Materials property part of my input file, I can add constants L, kappa_op, mu, and gama_asymm to run the simulation. If I use the TotalFreeEngery type to output the free energy, it requires me to add the free energy function. But the GrainGrowth action should already have the free energy function. I do not know how can I use it. Is there a way to output the free energy? Attached is my input file. 

Thanks very much.

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 50
  ny = 50
  nz = 0
  xmin = 0
  xmax = 100
  ymin = 0
  ymax = 100
  zmin = 0
  zmax = 0
  elem_type = QUAD4
[]

[GlobalParams]
  op_num = 3
  var_name_base = gr
[]

[Variables]
  [./PolycrystalVariables]
  [../]
[]

[UserObjects]
  [./voronoi]
    type = PolycrystalVoronoi
    rand_seed = 105
    grain_num = 3
    coloring_algorithm = bt
  [../]
[]

[ICs]
  [./PolycrystalICs]
    [./PolycrystalColoringIC]
      polycrystal_ic_uo = voronoi
    [../]
  [../]
[]

[AuxVariables]
  [./bnds]
    order = FIRST
    family = LAGRANGE
  [../]
  
  [./local_free_energy]
    family = MONOMIAL
    order = CONSTANT
  [../]
  
[]


[Modules]
  [./PhaseField]
    [./GrainGrowth]
      variable_mobility = false
      kappa = kappa_op
      mobility = L           
    [../]
  [../]
[]

[AuxKernels]
  [./BndsCalc]
    type = BndsCalcAux
    variable = bnds
    execute_on = timestep_end
  [../]

  [./local_free_energy]
    type = TotalFreeEnergy
    variable = local_free_energy
    f_name = F
    kappa_names = 'kappa_op kappa_op kappa_op'
    interfacial_vars ='gr0 gr1 gr2'   
  [../]

[]

#[BCs]
  #[./Periodic]
    #[./All]
     # auto_direction = 'x y'
    #[../]
 # [../]
#[]


[Materials]
  [./constant]
    type = GenericConstantMaterial
    prop_names  = 'L     kappa_op  mu   gamma_asymm'
    prop_values = '1.0    2.0       1.0   1.5' 
  [../]  
  [./free_energy]
   type = DerivativeParsedMaterial
   f_name = F                     
   args = 'gr2 gr1 gr0'           
   outputs = exodus
  [../] 

[]

[Postprocessors]
  active = ''
  [./ngrains]
    type = FeatureFloodCount
    variable = bnds
    threshold = 0.7
  [../]

  [./total_free_energy]
    type = ElementIntegralVariablePostprocessor
    variable = local_free_energy
  [../]
  
  [./total_gr0]
    type = ElementIntegralVariablePostprocessor
    variable = gr0
  [../]
  
  [./total_gr1]
    type = ElementIntegralVariablePostprocessor
    variable = gr1
  [../]

  [./total_gr2]
    type = ElementIntegralVariablePostprocessor
    variable = gr2
  [../]  
[]

[Preconditioning]
  active = ''
  [./SMP]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  scheme = 'bdf2'
  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type -ksp_gmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm        31    preonly    ilu          1' 

  l_tol = 1.0e-4
  l_max_its = 30
  nl_max_its = 20
  nl_rel_tol = 1.0e-9
  start_time = 0.0
  num_steps = 3
  dt = 1
[]

[Outputs]
  exodus = true
  csv = true
  perf_graph = true
  interval = 1
[]
```

**Comment by dschwen** [comment, marked as answer]
If you set `f_name = 0` you will get only the interfacial contributions, which should be all you need.

**Reply by dschwen** [reply]
Actually, strike that. That definitely will not be enough! You could implement a Material class to compute the local free energy density, according to the equation for _f<sub>loc</sub>_ in https://mooseframework.inl.gov/moose/modules/phase_field/Grain_Growth_Model.html

**Reply by dschwen** [reply]
Or derive from `TotalFreeEnergyBase`...

**Reply by Clark-domain** [reply]
Thanks very much! I think I got your idea. Total free energy should include floc and fgrd. I wanted to use the existing MOOSE objects, so I added the expression for free energy function F in the Materials object (see below), and kept using the TotalFreeEnergy object.  Now I got the outputs from the .e file and was able to see it from Varaview. But I have a new problem that I could not output the .csv file though I set csv = true. So I have no information on total free energy and grain areas for me to plot out in Excel. Do I still have something else wrong with the input file? 

[./free_energy]
   type = DerivativeParsedMaterial
   f_name = F                     
   args = 'gr2 gr1 gr0'           
   function = 'mu*((0.25*gr2^4-0.5*gr2^2)
+(0.25*gr1^4-0.5*gr1^2)
+(0.25*gr0^4-0.5*gr0^2)
+gamma_asymm*((gr2^2)*(gr1^2)
+(gr2^2)*(gr0^2)
+(gr1^2)*(gr0^2))+0.25)'        
   derivative_order = 2
   material_property_names = 'mu gamma_asymm'
   outputs = exodus
  [../]

**Reply by dschwen** [reply]
Hm, if you have `Postrocessors` and `Outputs/csv=true`, you should definitely get a CSV output file with the PP values in it.

**Reply by Clark-domain** [reply]
Problem solved. I checked it again, and I realized that I should not use active = ' ' in the Postprocessors,  or I should set active for the outputs I wanted. Thanks very much.
